File: Lib/jxxload_help/PathVFS.java
// Copyright 2000 Samuele Pedroni

package jxxload_help;

import java.util.Vector;
import java.util.Hashtable;
import java.util.Enumeration;
import java.util.zip.ZipFile;
import java.util.zip.ZipEntry;
import java.io.*;

public class PathVFS extends Object {
    
    public interface VFS {
        
       public InputStream open(String id);
        
    }
        
    public static class JarVFS implements VFS {
        private ZipFile zipfile;
        
        public JarVFS(String fname) throws IOException {
            zipfile = new ZipFile(fname);
        }
        
        public InputStream open(String id) {
            ZipEntry ent = zipfile.getEntry(id);
            if (ent == null) return null;
            try {
                return zipfile.getInputStream(ent);
            } catch(IOException e) {
                return null;
            }
        }
        
    }
    
    public static class DirVFS implements VFS {
        private String prefix;
        
        public DirVFS(String dir) {
            if (dir.length() == 0)
                prefix = null;
            else
                prefix = dir;
        }
        
        public InputStream open(String id) {
            File file = new File(prefix,id.replace('/',File.separatorChar));
            if (file.isFile()) {
                try {
                    return new BufferedInputStream(new FileInputStream(file));
                } catch(IOException e) {
                    return null;
                }
            }           
            return null;
        }
    }
    
    private Vector vfs = new Vector();
    private Hashtable once = new Hashtable();
    
    private final static Object PRESENT = new Object();
    
    public void addVFS(String fname) {
        if (fname.length() == 0) {
            if (!once.containsKey("")) {
                once.put("",PRESENT);
                vfs.addElement(new DirVFS(""));
            }
            return;
        }
        try {
            File file = new File(fname);
            String canon = file.getCanonicalPath().toString();
            if (!once.containsKey(canon)) {
                once.put(canon,PRESENT);
                if (file.isDirectory()) vfs.addElement(new DirVFS(fname));
                else if (file.exists() &&  (fname.endsWith(".jar") || fname.endsWith(".zip"))) {
                    vfs.addElement(new JarVFS(fname));
                }
            }
            
        } catch(IOException e) {}
    }
    
    public InputStream open(String id) {
        for(Enumeration enum = vfs.elements(); enum.hasMoreElements();) {
            VFS v = (VFS)enum.nextElement();
            InputStream stream = v.open(id);
            if (stream != null) return stream;
        }
        return null;
    }
      
}

File: Tools/jythonc/PathVFS.py
# Copyright 2000 Samuele Pedroni
import sys

from java import io
from java.util import zip


class JarVFS:
    def __init__(self,fname):
        self.zipfile = zip.ZipFile(fname)

    def open(self,id):
        ent = self.zipfile.getEntry(id)
        if ent:
##            print "path-jar-open: %s!%s" % (self.zipfile.name,id) # ?? dbg
            return self.zipfile.getInputStream(ent)
        else:
            return None

    def __repr__(self):
        return "<jar-vfs '%s'>" % self.zipfile.name
        
class DirVFS:
    def __init__(self,dir):
        if dir == '':
            self.pfx = None
        else:
            self.pfx = dir

    def open(self,id):
        f = io.File(self.pfx, id.replace('/',io.File.separator))
        if f.file:
##            print "path-open:",f # ?? dbg
            return io.BufferedInputStream(io.FileInputStream(f))
        return None

    def __repr__(self):
        return "<dir-vfs '%s'>" % self.pfx

class PathVFS:

    def add_vfs(self,fname):
        if fname == '':
            if not self.once.has_key(''):
                self.once['']=1
                self.vfs.append(DirVFS(''));
            return        
        file=io.File(fname);
        canon = file.canonicalPath
        if not self.once.has_key(canon):
            self.once[canon]=1
            try:
                if file.directory:
                    self.vfs.append(DirVFS(fname));
                else:
                    if file.exists and (fname.endswith('.jar') or fname.endswith('.zip')):
                        self.vfs.append(JarVFS(fname))
            except:
                pass
        

    def __init__(self,registry):
        self.once = {}
        self.vfs = []
        paths = registry.getProperty("python.packages.paths","java.class.path")
        paths = paths.split(',')
        # opt
        if "sun.boot.class.path" in paths: # ??pending strip boot class paths of other jvms?
            paths.remove("sun.boot.class.path")

        #paths.append("python.packages.fakepath")

        for p in paths:
            e = registry.getProperty(p)
            if e != None:
                path = e.split(io.File.pathSeparator)
                for name in path:
                    self.add_vfs(name)

        for name in sys.path:
            self.add_vfs(name)
            
        del self.once

    def open(self,id):
        for v in self.vfs:
            stream = v.open(id)
            if stream:
                return stream
        return None
